monte_carlo.py

```diff
@@ -0,0 +1,195 @@
+            self.policy.set_trained()
+
+    @staticmethod
+    def get_legal_actions(state):
+        candidates = MonteCarlo.actions.values()
+        legal_actions = []
+        for c in candidates:
+            combined = (state[2] + c[0], state[3] + c[1])
+            if np.all(np.abs(combined) <= MonteCarlo.max_velocity) and np.any(combined):
+                legal_actions.append(c)
+
+        legal_actions = [list(MonteCarlo.actions.keys())[list(MonteCarlo.actions.values()).index(a)] for a in legal_actions]
+        return legal_actions
+    
+    def generate_episode(self, start=None):
+        
+        episode = []
+        if start is None:
+            random_start_index = np.random.choice(len(self.starts))
+            random_start = self.starts[random_start_index]
+            start = random_start
+        state = start + (0, 0)
+        while state[:2] != self.goal:
+            action = self.policy[state]
+            dvx, dvy = MonteCarlo.actions[action]
+            vx, vy = state[2] + dvx, state[3] + dvy
+            # Ensure velocity stays within bounds
+            vx = max(-self.max_velocity, min(self.max_velocity, vx))
+            vy = max(-self.max_velocity, min(self.max_velocity, vy))
+
+            next_position = (state[0] + vx, state[1] + vy)
+            next_state = next_position + (vx, vy)
+
+            if next_position[0] < 0 or next_position[0] >= self.grid.shape[0] or next_position[1] < 0 or next_position[1] >= self.grid.shape[1] or self.hits_wall(state[:2], next_position):
+                reward = -1  # Penalty for hitting boundaries or obstacles
+                next_state = start + (0, 0)
+            elif next_position == self.goal:
+                reward = 1
+            else:
+                reward = -1  # Small penalty for each move
+            
+            episode.append((state, action, reward))
+            state = next_state
+        return episode
+    
+    def hits_wall(self, pos1, pos2):
+        """
+        Find if the line between pos1 and pos2 hits a wall.
+        """
+        x1, y1 = pos1
+        x2, y2 = pos2
+        dx = abs(x2 - x1)
+        dy = abs(y2 - y1)
+        x = x1
+        y = y1
+        n = 1 + dx + dy
+        x_inc = 1 if x2 > x1 else -1
+        y_inc = 1 if y2 > y1 else -1
+        error = dx - dy
+        dx *= 2
+        dy *= 2
+        for _ in range(n):
+            if self.grid[x, y] == CELL.WALL:
+                return True
+            if error > 0:
+                x += x_inc
+                error -= dy
+            else:
+                y += y_inc
+                error += dx
+        return False
+
+    def monte_carlo_control(self):
+        if self.is_trained:
+            warnings.warn("The agent is already trained.")
+            return
+        
+        for _ in tqdm(range(self.num_episodes)):
+            episode = self.generate_episode()
+            G = 0
+            for t in reversed(range(len(episode))):
```

Comment:
isn't simply a for loop from back to front more efficient? just a small thing

Reply:
Please explain more what you mean. The top for loop is simply over the number of episodes and the inner for loop is already from back to front.

Reply:
instead of using the reversed method we could do something like: `for t in range(len(episode) - 1, -1, -1)`